### PR TITLE
Method ambiguity resolution

### PR DESCRIPTION
**Pull request by vlad20012**
All `RsPreciseTraitMatchingTest` tests now passes
1. We now filter methods of traits that are not imported
2. We now check trait bounds of impls

We do not do anything above if we don't have method name ambiguity, i.e. name resolution succeeded with exactly single entity

### CODE REVIEW

**Comment by vlad20012** [comment]
About the last commit: The idea is that if we resolved exactly 1 method, we don't perform
any additional filtering and just use this 1 method.

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 1]
```diff
@@ -472,6 +475,12 @@ class ImplLookup(
             return TyFunction(argumentTypes, outputType)
         }
 
+    fun isTraitVisibleFrom(trait: RsTraitItem, scope: RsElement): Boolean {
+        val name = trait.name ?: return true
+        val path = RsPsiFactory(project).tryCreatePath(name)?.apply { setContext(scope) } ?: return true
+        return resolvePath(path, this).any { it.element == trait }
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: I wonder if these uncaheble resolves have any impact on performance.

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: This code anyway executes very rarely - only if we run into method name ambiguity (resolved 2 methods with the same name). I've checked it with our `the best` :) performance tests and did not noticed any regression

Comment (matklad) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: Yeah, a fair point about this being rare!

I foresee though that one day we'll start to use this for "import trait" quick fix, and it'll become a bottleneck :) But we definitely should not try to optimize it right now!

**Pull request merged**